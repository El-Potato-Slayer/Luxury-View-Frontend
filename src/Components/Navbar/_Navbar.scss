.navbar {
  display: grid;
  grid-template-rows: 1fr 2.75fr;
  position: fixed;
  top: 0;
  left: -14rem;
  width: 14rem;
  height: 100%;
  padding-left: 1rem;
  z-index: 3;
  background: $white;
  border-right: 1px solid $light-gray;
  transition: all 0.75s ease-in-out;
}

.navbar.open {
  left: 0;
}

.dark-filter {
  position: absolute;
  top: 0;
  left: calc(14rem);
  width: calc(100vw);
  height: 100%;
  background: rgba(0, 0, 0, 0.5);
  z-index: 0;
}

.hide {
  display: none;
}

.nav-logo {
  margin: 0 1rem;
  width: 70%;
}

.nav-links {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.nav-item {
  padding: 0.75rem 1rem;
  width: 100%;
}

.nav-item:hover {
  background-color: $green;
  color: $white;
}
