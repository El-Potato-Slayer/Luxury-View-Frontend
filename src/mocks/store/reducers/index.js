import { combineReducers } from 'redux';
import agentsReducer from './agentsReducer';
import appointmentsReducer from './appointmentsReducer';
import authReducer from './authReducer';
import propertiesReducer from './propertiesReducer';
import userReducer from './userReducer';
import filterReducer from './filterReducer';

export default combineReducers({
  agentsReducer, appointmentsReducer, authReducer, propertiesReducer, userReducer, filterReducer,
});
