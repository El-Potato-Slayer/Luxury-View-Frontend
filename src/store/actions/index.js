export const setAgents = (payload) => ({
  type: 'SET_AGENTS',
  payload,
});

export const setAppointments = (payload) => ({
  type: 'SET_APPOINTMENTS',
  payload,
});

export const setProperties = (payload) => ({
  type: 'SET_PROPERTIES',
  payload,
});

export const changeFilter = (name) => ({
  type: 'CHANGE_FILTER',
  name,
});
