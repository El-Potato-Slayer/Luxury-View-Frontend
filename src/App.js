import './App.css';
import {
  BrowserRouter, Redirect, Route, Switch,
} from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Navbar from './components/Navbar';
import Home from './components/Home';
import { setAgents, setAuth, setProperties } from './store/actions';
import MansionsList from './containers/MansionsList';
import AgentsList from './containers/AgentsList';
import ProtectedRoute from './components/ProtectedRoute';
import AppointmentsList from './containers/AppointmentsList';
import Login from './components/Login';
import Mansion from './components/Mansion';

function App({ isAuth }) {
  const dispatch = useDispatch();
  let token;
  useEffect(() => {
    axios.get('http://localhost:3000/api/v1/properties')
      .then((resp) => {
        dispatch(setProperties(resp.data));
      });
    axios.get('http://localhost:3000/api/v1/agents')
      .then((resp) => {
        dispatch(setAgents(resp.data));
      });
    token = localStorage.getItem('token');
    if (token) {
      const tokenExpiration = JSON.parse(atob(token.split('.')[1])).exp;
      const dateNow = new Date();
      if (tokenExpiration > dateNow.getTime() / 1000) {
        dispatch(setAuth(true));
      }
    }
  }, []);
  return (
    <div data-testid="app" className="App">
      <BrowserRouter>
        <Navbar />
        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route exact path="/mansions">
            <MansionsList />
          </Route>
          <Route exact path="/agents">
            <AgentsList />
          </Route>
          <Route path="/login">
            {isAuth ? <Redirect to="/appointments" /> : <Login />}
          </Route>
          <Route path="/mansions/:id">
            <Mansion />
          </Route>
          <ProtectedRoute exact path="/appointments" component={AppointmentsList} />
        </Switch>
      </BrowserRouter>
    </div>
  );
}

const mapStateToProps = (state) => ({
  isAuth: state.authReducer.isAuth,
});

App.propTypes = {
  isAuth: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps)(App);
