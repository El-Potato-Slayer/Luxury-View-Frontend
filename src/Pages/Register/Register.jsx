import axios from 'axios';
import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { setUser } from '../../Redux/actions/userActions';

function Register() {
  const dispatch = useDispatch();
  const history = useHistory();

  const [input, setInput] = useState({
    username: '',
    email: '',
    password: '',
    first_name: '',
    last_name: '',
  });
  const [error, setError] = useState('');

  const inputHandler = (e) => {
    setInput({ ...input, [e.target.name]: e.target.value });
  };

  function displayError(error) {
    if (error) {
      return (
        <p>{error}</p>
      );
    }
    return null;
  }

  const registerUser = (e) => {
    e.preventDefault();
    const user = { user: { ...input } };
    axios.post('register', user)
      .then((res) => {
        localStorage.setItem('token', res.data.token);
        dispatch(setUser(res.data));
        history.push('/');
      })
      .catch(() => {
        setError('An error occurred. Please try again later');
      });
  };

  return (
    <div data-testid="register" className="form-wrapper">
      {displayError(error)}
      <form className="form">
        <h2 className="page-title">Register</h2>
        <fieldset>
          <input className="input-field" type="text" placeholder="First Name" name="first_name" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input className="input-field" type="text" placeholder="Last Name" name="last_name" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input className="input-field" type="text" placeholder="Username" name="username" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input className="input-field" type="text" placeholder="Email" name="email" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input className="input-field" type="password" placeholder="Password" name="password" onChange={inputHandler} />
        </fieldset>
        <button className="button info" type="submit" onClick={registerUser}>Submit</button>
      </form>
    </div>
  );
}

export default Register;
