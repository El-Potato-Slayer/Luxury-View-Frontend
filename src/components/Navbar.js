import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { clearUser } from '../store/actions/userActions';
// import { setAuth, setUser } from '../store/actions';

function Navbar() {
  const [burgerState, setBurgerState] = useState(false);
  const { isLoggedIn } = useSelector((state) => state.userReducer);
  const dispatch = useDispatch();
  const logout = () => {
    localStorage.clear();
    dispatch(clearUser());
  };

  const handleBurgerState = () => {
    setBurgerState(!burgerState);
  };

  function authNavItem(isLoggedIn) {
    if (isLoggedIn) {
      return (
        <Link onClick={() => { logout(); handleBurgerState(); }} data-testid="navbar-logout" className="nav-item" to="/">Logout</Link>
      );
    }
    return (
      <>
        <Link onClick={handleBurgerState} data-testid="navbar-login" className="nav-item" to="/login">Login</Link>
        <Link onClick={handleBurgerState} data-testid="navbar-register" className="nav-item" to="/register">Register</Link>
      </>
    );
  }

  return (
    <>
      <div data-testid="navbar" className={`navbar${burgerState ? ' open' : ''}`}>
        <div aria-hidden="true" onClick={handleBurgerState} onKeyDown={handleBurgerState} className={`dark-filter${!burgerState ? ' hide' : ''}`} />
        <div className={`burger${burgerState ? ' open' : ''}`}>
          <button type="button" id="burger-wrapper" className="burger-wrapper" onClick={handleBurgerState}>
            <div id="burger-menu" className={`burger-menu${burgerState ? ' open' : ''}`} />
          </button>
        </div>
        <img src="/assets/logo.png" className="nav-logo" alt="" />
        <div className="nav-links">
          <Link onClick={handleBurgerState} data-testid="navbar-homepage" className="nav-item" to="/">Home</Link>
          <Link onClick={handleBurgerState} data-testid="navbar-mansions" className="nav-item" to="/mansions">Mansions</Link>
          <Link onClick={handleBurgerState} data-testid="navbar-appointments" className="nav-item" to="/appointments">Appointments</Link>
          <Link onClick={handleBurgerState} data-testid="navbar-agents" className="nav-item" to="/agents">Agents</Link>
          {authNavItem(isLoggedIn)}
        </div>
        <div className="nav-social" />
      </div>
    </>
  );
}

export default Navbar;
