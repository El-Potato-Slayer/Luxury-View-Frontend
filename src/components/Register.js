import axios from 'axios';
import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { setAuth, setUser } from '../store/actions';

function Register() {
  const dispatch = useDispatch();

  const [input, setInput] = useState({
    username: '',
    email: '',
    password: '',
    first_name: '',
    last_name: '',
  });
  const [error, setError] = useState('');

  const inputHandler = (e) => {
    setInput({ ...input, [e.target.name]: e.target.value });
    console.log(input);
  };

  function displayError(error) {
    if (error) {
      return (
        <p>{Error}</p>
      );
    }
    return null;
  }

  const registerUser = (e) => {
    e.preventDefault();
    const user = { user: { ...input } };
    axios.post('register', user)
      .then((res) => {
        localStorage.setItem('token', res.data.token);
        dispatch(setUser(res.data));
        dispatch(setAuth(true));
        console.log(res.data);
      })
      .catch(() => {
        setError('Username or password is incorrect');
      });
  };

  return (
    <>
      <h2>Register</h2>
      {displayError(error)}
      <form action="">
        <fieldset>
          <input type="text" placeholder="First Name" name="first_name" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input type="text" placeholder="Last Name" name="last_name" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input type="text" placeholder="Username" name="username" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input type="text" placeholder="Email" name="email" onChange={inputHandler} />
        </fieldset>
        <fieldset>
          <input type="text" placeholder="Password" name="password" onChange={inputHandler} />
        </fieldset>
        <button type="submit" onClick={registerUser}>Submit</button>
      </form>
    </>
  );
}

export default Register;
