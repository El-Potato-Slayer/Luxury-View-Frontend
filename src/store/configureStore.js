import { applyMiddleware, createStore } from 'redux';
import ReduxThunk from 'redux-thunk';
import { fetchData } from './actions';
import { fetchAgentsFailure, fetchAgentsRequest, fetchAgentsSuccess } from './actions/agentActions';
import { fetchAppointmentsFailure, fetchAppointmentsRequest, fetchAppointmentsSuccess } from './actions/appointmentActions';
import { fetchPropertiesFailure, fetchPropertiesRequest, fetchPropertiesSuccess } from './actions/propertyActions';
import rootReducer from './reducers';

const store = createStore(rootReducer, applyMiddleware(ReduxThunk));

// store.subscribe(() => { console.log(store.getState()); });
store.dispatch(fetchData('properties', fetchPropertiesRequest, fetchPropertiesSuccess, fetchPropertiesFailure));
store.dispatch(fetchData('agents', fetchAgentsRequest, fetchAgentsSuccess, fetchAgentsFailure));
store.dispatch(fetchData('appointments', fetchAppointmentsRequest, fetchAppointmentsSuccess, fetchAppointmentsFailure, true));

export default store;
