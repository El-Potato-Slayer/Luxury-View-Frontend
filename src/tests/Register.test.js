import {
  act,
  cleanup, fireEvent, render, screen, waitFor,
} from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import store from '../mocks/store/configureStore';
import App from '../App';
import Register from '../components/Register';
import Navbar from '../components/Navbar';

jest.mock('axios');
afterEach(cleanup);

test('Renders the Register component without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Register />
        </BrowserRouter>
      </App>
    </Provider>, div,
  );
});

test('Checks that title is rendered', async () => {
  act(() => {
    render(
      <Provider store={store}>
        <App>
          <BrowserRouter>
            <Navbar />
          </BrowserRouter>
        </App>
      </Provider>,
    );
  });
  act(() => {
    fireEvent.click(screen.getByTestId('navbar-register'));
  });

  await waitFor(() => {
    expect(screen.getByTestId('register')).toHaveTextContent('Register');
  });
});

test('Checks that unknown title is not rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.getByTestId('register')).not.toHaveTextContent('Jakey');
});

test('Checks that first name input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('First Name')).toBeInTheDocument();
});

test('Checks that last name input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Last Name')).toBeInTheDocument();
});

test('Checks that username input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Username')).toBeInTheDocument();
});

test('Checks that email input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Email')).toBeInTheDocument();
});

test('Checks that password input is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Password')).toBeInTheDocument();
});

test('Checks that unknown input is not rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByPlaceholderText('Quake')).not.toBeInTheDocument();
});

test('Checks that submit button is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.queryByText('Submit')).toBeInTheDocument();
});

it('matches snapshot', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <BrowserRouter>
        <Register />
      </BrowserRouter>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
