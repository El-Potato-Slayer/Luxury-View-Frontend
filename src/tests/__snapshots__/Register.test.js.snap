// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="form-wrapper"
  data-testid="register"
>
  <form
    className="form"
  >
    <h2
      className="page-title"
    >
      Register
    </h2>
    <fieldset>
      <input
        className="input-field"
        name="first_name"
        onChange={[Function]}
        placeholder="First Name"
        type="text"
      />
    </fieldset>
    <fieldset>
      <input
        className="input-field"
        name="last_name"
        onChange={[Function]}
        placeholder="Last Name"
        type="text"
      />
    </fieldset>
    <fieldset>
      <input
        className="input-field"
        name="username"
        onChange={[Function]}
        placeholder="Username"
        type="text"
      />
    </fieldset>
    <fieldset>
      <input
        className="input-field"
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="text"
      />
    </fieldset>
    <fieldset>
      <input
        className="input-field"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="text"
      />
    </fieldset>
    <button
      className="info-button"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
